src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.DecryptToken.cs

```diff
@@ -0,0 +1,270 @@
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            TokenLogMessages.IDX10000,
+                            LogHelper.MarkAsNonPII(nameof(jwtToken))),
+                        typeof(ArgumentNullException),
+                        new System.Diagnostics.StackFrame()));
+
+            if (validationParameters == null)
+                return new TokenDecryptionResult(
+                    jwtToken,
+                    ValidationFailureType.TokenDecryptionFailed,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            TokenLogMessages.IDX10000,
+                            LogHelper.MarkAsNonPII(nameof(validationParameters))),
+                        typeof(ArgumentNullException),
+                        new System.Diagnostics.StackFrame()));
+
+            if (string.IsNullOrEmpty(jwtToken.Enc))
+                return new TokenDecryptionResult(
+                    jwtToken,
+                    ValidationFailureType.TokenDecryptionFailed,
+                    new ExceptionDetail(
+                        new MessageDetail(TokenLogMessages.IDX10612),
+                        typeof(SecurityTokenException),
+                        new System.Diagnostics.StackFrame()));
+
+            var keysOrExceptionDetail = GetContentEncryptionKeys(jwtToken, validationParameters, configuration, callContext);
+            if (keysOrExceptionDetail.Item2 != null) // ExceptionDetail returned
+                return new TokenDecryptionResult(
+                    jwtToken,
+                    ValidationFailureType.TokenDecryptionFailed,
+                    keysOrExceptionDetail.Item2);
+
+            var keys = keysOrExceptionDetail.Item1;
+            if (keys == null)
+                return new TokenDecryptionResult(
+                    jwtToken,
+                    ValidationFailureType.TokenDecryptionFailed,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            TokenLogMessages.IDX10609,
+                            LogHelper.MarkAsSecurityArtifact(jwtToken, JwtTokenUtilities.SafeLogJwtToken)),
+                        typeof(SecurityTokenDecryptionFailedException),
+                        new System.Diagnostics.StackFrame()));
+
+            return JwtTokenUtilities.DecryptJwtToken(
+                jwtToken,
+                validationParameters,
+                new JwtTokenDecryptionParameters
+                {
+                    DecompressionFunction = JwtTokenUtilities.DecompressToken,
+                    Keys = keys,
+                    MaximumDeflateSize = MaximumTokenSizeInBytes
+                },
+                callContext);
+        }
+
+        internal (IList<SecurityKey>?, ExceptionDetail?) GetContentEncryptionKeys(JsonWebToken jwtToken, ValidationParameters validationParameters, BaseConfiguration configuration, CallContext? callContext)
+        {
```

Comment:
this method is quite large...can it be broken into smaller helper methods, which are also easier to test?